File: backend/app/models/customer.py
"""Customer Entity ORM Models.

Defines the Customer entity representing client accounts, including health score metrics,
lifecycle statuses (ACTIVE, AT_RISK, CHURNED, PROSPECT), and assigned CSM account owners.
"""

import enum
import uuid
from typing import List, Optional, TYPE_CHECKING
from sqlalchemy import CheckConstraint, Enum, ForeignKey, Integer, String, Text, Uuid
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database.base import Base, TimestampMixin, UUIDPrimaryKeyMixin

if TYPE_CHECKING:
    from app.models.user import User
    from app.models.interaction import Interaction


class CustomerStatus(str, enum.Enum):
    """Enumeration of customer account lifecycle states."""

    ACTIVE = "ACTIVE"
    AT_RISK = "AT_RISK"
    CHURNED = "CHURNED"
    PROSPECT = "PROSPECT"


class Customer(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """Customer database model representing client accounts.

    Attributes:
        id: Unique UUID identifier.
        name: Primary contact person's name.
        company_name: Name of the customer organization.
        email: Primary contact email address.
        phone: Optional contact telephone number.
        industry: Business industry classification.
        status: Lifecycle status of the customer (ACTIVE, AT_RISK, CHURNED, PROSPECT).
        health_score: Composite retention/satisfaction score (0-100).
        owner_id: UUID of assigned CSM / Account Owner.
        notes: General account observations or remarks.
        owner: User relationship representing the account manager.
        interactions: List of historical interactions associated with this customer.
        created_at: UTC timestamp of record creation.
        updated_at: UTC timestamp of last record update.
    """

    __tablename__ = "customers"
    __table_args__ = (
        CheckConstraint("health_score >= 0 AND health_score <= 100", name="check_health_score_range"),
    )

    name: Mapped[str] = mapped_column(String(150), index=True, nullable=False)
    company_name: Mapped[str] = mapped_column(String(150), index=True, nullable=False)
    email: Mapped[str] = mapped_column(String(255), nullable=False)
    phone: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    industry: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    status: Mapped[CustomerStatus] = mapped_column(
        Enum(CustomerStatus, name="customer_status", create_type=False),
        default=CustomerStatus.ACTIVE,
        index=True,
        nullable=False,
    )
    health_score: Mapped[int] = mapped_column(
        Integer, default=100, index=True, nullable=False
    )
    owner_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        Uuid(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        index=True,
        nullable=True,
    )
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    # Relationships
    owner: Mapped[Optional["User"]] = relationship("User", back_populates="customers")
    interactions: Mapped[List["Interaction"]] = relationship(
        "Interaction",
        back_populates="customer",
        cascade="all, delete-orphan",
        order_by="desc(Interaction.meeting_date)",
    )

    def __repr__(self) -> str:
        """Return developer-friendly string representation of the Customer instance."""
        return f"<Customer id={self.id} company={self.company_name} status={self.status} health={self.health_score}>"

File: frontend/types/customer.ts
/**
 * Customer Account Types.
 *
 * Defines customer lifecycle statuses, Customer ORM representations, and query filter interfaces.
 */

import { User } from './auth';
import { BaseQueryParams } from './common';

/** Customer lifecycle classification */
export type CustomerStatus = 'ACTIVE' | 'AT_RISK' | 'CHURNED' | 'PROSPECT';

/** Customer account record */
export interface Customer {
  /** Unique UUID identifier */
  id: string;
  /** Primary contact person name */
  name: string;
  /** Company organization name */
  company_name: string;
  /** Contact email */
  email: string;
  /** Optional phone number */
  phone?: string | null;
  /** Business industry */
  industry?: string | null;
  /** Account lifecycle status */
  status: CustomerStatus;
  /** Numeric health score rating (0-100) */
  health_score: number;
  /** UUID of assigned CSM / Owner */
  owner_id?: string | null;
  /** Populated CSM User profile */
  owner?: User | null;
  /** Internal account notes */
  notes?: string | null;
  /** Record creation timestamp */
  created_at: string;
  /** Last modification timestamp */
  updated_at: string;
}

/** Query parameters for customer listing and filtering */
export interface CustomerQueryParams extends BaseQueryParams {
  /** Text search across name, company, email */
  search?: string;
  /** Status filter */
  status?: CustomerStatus;
  /** Lower health score limit */
  min_health_score?: number;
  /** Upper health score limit */
  max_health_score?: number;
  /** CSM owner UUID filter */
  owner_id?: string;
  /** Industry category filter */
  industry?: string;
}

